_posts/2020-08-08-a-look-at-the-aspnet-core-logging-provider-for-app-service.md

```diff
@@ -0,0 +1,532 @@
+        {
+            namedSetup.Configure(name, options);
+        }
+        else if (name == Options.DefaultName)
+        {
+            setup.Configure(options);
+        }
+    }
+
+    // 3. Finally, run all post-configuration steps.
+    foreach (var post in _postConfigures)
+    {
+        post.PostConfigure(name, options);
+    }
+
+    // 4. Omitted for brevity is the execution of the validation steps.
+    [...]
+
+    // 5. Return the configured options value.
+    return options;
+}
+```
+
+### How the logging system is aware of options changes
+
+The logging system supports changes to the logging options by internally using an instance of `IOptionsMonitor<LoggerFilterOptions>`, and subscribing to its changes, triggering a reevaluation of which filters need to be applied.
+Without going into the filter selection logic, this piece of code from the `LoggerFactory` class reveals how it subscribes to changes:
+
+```csharp
+public LoggerFactory(
+    IEnumerable<ILoggerProvider> providers,
+    IOptionsMonitor<LoggerFilterOptions> filterOption)
+{
+    foreach (var provider in providers)
+    {
+        AddProviderRegistration(provider, dispose: false);
+    }
+
+    // 1. Subscribe to options changes, and trigger a refresh of filters when one is triggered.
+    _changeTokenRegistration = filterOption.OnChange(RefreshFilters);
+
+    // 2. Set up filters according to the initial value of logging options.
+    RefreshFilters(filterOption.CurrentValue);
+}
+
+private void RefreshFilters(LoggerFilterOptions filterOptions)
+{
+    // Implementation omitted.
+    // Refreshes filters applied to existing logging providers.
+}
+```
+
+### Putting it all together
+
+We now have all the pieces to understand how it all works, as we have established the following sequence of events:
+
+1. In the Azure portal or through the Azure CLI, we make a change to the logging configuration on our App Service instance;
+1. The new configuration settings are persisted in a JSON file on the App Service file system;
+1. The configuration system loads up the new settings from the updated file and triggers a change notification;
+1. This configuration change notification is also picked up by the options system through the configuration-to-options adapter, which triggers a rebuild of the `LoggerFilterOptions` value;
+1. The options system executes all the registered configuration actions for `LoggerFilterOptions`;
+1. One of these configuration actions, as seen earlier, adds a logging filtering rule for the file system logging provider, using the minimum log level option configured in step 1;
+1. The options system notifies change subscribers about the new constructed value;
+1. The logging stack handles that notification and refreshes its filters; and finally
+1. The filter that targets the App Service file system logging provider ensures only entries with a matching log level get written to disk.
+
+Phew.
+We made it 🎉!
+Or did we?
```

Comment:
Won't markdown render these as a single line? Do you need double line breaks?

Reply:
No, they will be on a single line, which is what I was going for.

See the rendered version at <https://github.com/mderriey/mderriey.github.io/blob/add-post-about-aspnet-core-logging-provider-for-app-service/_posts/2020-08-08-a-look-at-the-aspnet-core-logging-provider-for-app-service.md#how-to-customise-it>